pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java

```diff
@@ -311,15 +307,26 @@ void updatePermitsAndPendingAcks(final List<Entry> entries, SendMessageInfo sent
         while (iter.hasNext()) {
             Entry entry = iter.next();
             ByteBuf metadataAndPayload = entry.getDataBuffer();
-            int batchSize = getBatchSizeforEntry(metadataAndPayload, subscription, consumerId);
-            if (batchSize == -1) {
-                // this would suggest that the message might have been corrupted
+            MessageMetadata msgMetadata = peekMessageMetadata(metadataAndPayload, subscription, consumerId);
+            PositionImpl pos = (PositionImpl) entry.getPosition();
+            if (msgMetadata == null) {
+                // Message metadata was corrupted
                 iter.remove();
-                PositionImpl pos = (PositionImpl) entry.getPosition();
                 entry.release();
                 subscription.acknowledgeMessage(Collections.singletonList(pos), AckType.Individual, Collections.emptyMap());
                 continue;
+            } else if (msgMetadata.hasDeliverAtTime()
```

Comment:
This is strange. The dispatcher gets a list of entries from the managed ledger, passes it to the consumer which then passes it back to an interface on the dispatcher which for all but one case is a noop.

It's also weird that tracking happens in a method called updatePermitsAndPendingAcks. And this method doesn't look like it should mutate it's argument, but it does.

The only reason I can see to do this is that you're trying to only call peekMessageMetadata once. However, the mutation means you are having to copy the list before the call to sendMessages, which is creating extra GC pressure even when tracking is disabled.

I think it would be better to contain it all in the dispatcher, and do the second peek. peekMessageMetadata is memory-neutral. So there would be more CPU usage but only in the case where delay is enabled. ```trackDelayedDelivery``` interface is no longer needed, as it can all be done before the call to sendMessages.

Something like:
```
readEntriesComplete(List<Entry> entries) {
    ...
    if (messagesForC > 0) {
        List<Entry> toSend = trackDelayedMessages(entries.subList(...));
        c.sendMessages(toSend);
    ...
}

private List<Entry> trackDelayedMessages(List<Entries> entries) {
    if (!trackingEnabled) {
        return entries;
    }
    // peek metadata for each entry
}
```

Reply:
> The only reason I can see to do this is that you're trying to only call peekMessageMetadata once. However, the mutation means you are having to copy the list before the call to sendMessages, which is creating extra GC pressure even when tracking is disabled.

The problem was already present before this change. If we fail the checksum or other error, the message is not pushed to consumer and removed from the list and would have cause the same modification error on the list.

Reply:
I can think 2 possible solutions to prevent modification error:

1. Have an intermediate inner list adding messages promoted to being discarded and call `entries.removeAll(entriesToDiscard);` outside the while.
2. Wrap entry list with `CopyOnWriteArrayList`, this will affect performance only when removing entries (by checksum validation or delay discard)

Reply:
> This is strange. The dispatcher gets a list of entries from the managed ledger, passes it to the consumer which then passes it back to an interface on the dispatcher which for all but one case is a noop.

As mentioned above, the list can be modified already if the checksum doesn't match or the metadata fails to deserialize. 

Let me think a bit more at how to refactor this.

> It's also weird that tracking happens in a method called updatePermitsAndPendingAcks. And this method doesn't look like it should mutate it's argument, but it does.

Agree, the method was already like that, and it could already modify the list before. Probably it's about time to rename it.

Reply:
> Agree, the method was already like that, and it could already modify the list before. Probably it's about time to rename it.

it would be nice to be able to do the peek before the send is called and have the result stored somewhere. Unfortunately I don't see a nice way to do so. 

Also, it's not just the mutation that I find weird. The data seems to be ping ponging between objects too.

Reply:
Yes. The thing is that if we look into the metadata in the dispatcher only, then we need a way to pass the batch size for each of the entries, or otherwise we have to deserialize the protobuf twice. 

The thing is that we would have to do the double de-serialization always (for shared subscription) since we wouldn't be able to know upfront if the message was marked with delay or not. 

Refactoring the `updatePermitsAndPendingAcks()` is a bit tricky because it's basically code that should belong to the dispatcher already but that it's in Consumer class since it's the same across all dispatchers. 

That's basically the context on why I went the route of checking the delay value in that particular position.

The refactoring of `updatePermitsAndPendingAcks()` would have to go in a different PR anyways.

Reply:
Yes, the tricky part is really that Consumer needs to track the batch size for each of the entries, so we cannot just get the total and pass that on to it.

Reply:
@sijie @ivankelly PTAL again, I have removed the list copy. Instead of the iterator.remove(), I changed it to just "null" the entries that are being removed. That works fine with the sublist.

I also renamed the updatePermitsAndPendingAcks to be more explicit in that it filters out messages (either for delay or for metadata/checksum errors). We can think later on how to refactor the interactions between dispatcher and consumers.